Marlin/src/module/planner.cpp

```diff
@@ -996,29 +998,13 @@ void Planner::calculate_trapezoid_for_block(block_t * const block, const_float_t
 
 // The kernel called by recalculate() when scanning the plan from last to first entry.
 void Planner::reverse_pass_kernel(block_t * const current, const block_t * const next, const_float_t safe_exit_speed_sqr) {
-  if (current) {
-    // If entry speed is already at the maximum entry speed, and there was no change of speed
-    // in the next block, there is no need to recheck. Block is cruising and there is no need to
-    // compute anything for this block,
-    // If not, block entry speed needs to be recalculated to ensure maximum possible planned speed.
-    const float max_entry_speed_sqr = current->max_entry_speed_sqr;
-
-    // Compute maximum entry speed decelerating over the current block from its exit speed.
-    // If not at the maximum entry speed, or the previous block entry speed changed
-    if (current->entry_speed_sqr != max_entry_speed_sqr || (next && next->flag.recalculate)) {
-
-      // If nominal length true, max junction speed is guaranteed to be reached.
-      // If a block can de/ac-celerate from nominal speed to zero within the length of the block, then
-      // the current block and next block junction speeds are guaranteed to always be at their maximum
-      // junction speeds in deceleration and acceleration, respectively. This is due to how the current
-      // block nominal speed limits both the current and next maximum junction speeds. Hence, in both
-      // the reverse and forward planners, the corresponding block junction speed will always be at the
-      // the maximum junction speed and may always be ignored for any speed reduction checks.
-
-      const float next_entry_speed_sqr = next ? next->entry_speed_sqr : safe_exit_speed_sqr,
-                  new_entry_speed_sqr = current->flag.nominal_length
-                    ? max_entry_speed_sqr
-                    : _MIN(max_entry_speed_sqr, max_allowable_speed_sqr(-current->acceleration, next_entry_speed_sqr, current->millimeters));
+  // We need to recalculate only for the last block added or if next->entry_speed_sqr changed.
+  if (!next || next->flag.recalculate) {
+    // And only if we're not already at max entry speed.
+    if (current->entry_speed_sqr != current->max_entry_speed_sqr) {
+      const float next_entry_speed_sqr = next ? next->entry_speed_sqr : safe_exit_speed_sqr;
+      float new_entry_speed_sqr = max_allowable_speed_sqr(-current->acceleration, next_entry_speed_sqr, current->millimeters);
+      NOMORE(new_entry_speed_sqr, current->max_entry_speed_sqr);
       if (current->entry_speed_sqr != new_entry_speed_sqr) {
```

Comment:
Am I right in thinking this if condition will never be false? `new_entry_speed_sqr` cannot equal to `current->max_entry_speed_sqr` (because the if statement above catches that) so it must be `max_allowable_speed_sqr(-current->acceleration, next_entry_speed_sqr, current->millimeters)`. This value will always be different to `current->entry_speed_sqr` if the `if (!next || next->flag.recalculate)` test succeeded.

Reply:
Technically, it's possible to be false on a segment that's next to last in the buffer. When this segment was previously added to the buffer, its entry speed got updated to `max_allowable_speed_sqr(-current->acceleration, safe_exit_speed_sqr, current->millimeters)`.
In this run of `reverse_pass_kernel()`, it will try updating to `max_allowable_speed_sqr(-current->acceleration, next_entry_speed_sqr, current->millimeters)` and `next_entry_speed_sqr` could by chance equal (an old) `safe_exit_speed_sqr` if there's a change to jerk/deviation settings which results in a lower new `safe_exit_speed_sqr` for the last segment in the buffer.

Very rare but also the check is also a very a simple comparison in pre-existing code. I could remove it if you want.

Reply:
>Technically, it's possible to be false on a segment that's next to last in the buffer...

Ah, yes.

> Very rare but also the check is a very simple comparison in pre-existing code. I could remove it if you want.

Let's leave it. It's minor and I don't want to create more need for testing.